plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Instance/Flat.hs

```diff
@@ -152,21 +151,25 @@ decodeTerm sizeLimit = go
             posPost <- getCurPtr
             let usedBytes = posPost `minusPtr` posPre
 
-            let conNode :: Term name uni fun ann
-                conNode = Constant ann con
+            let t :: Term name uni fun ann
+                t = Constant ann con
             case sizeLimit of
-                NoLimit -> pure conNode
-                Limit n ->
-                    if fromIntegral usedBytes > n
-                    then fail $ "Used more than " ++ show n ++ " bytes decoding the constant: " ++ show (prettyPlcDef conNode)
-                    else pure conNode
+                Limit n -> unless (fromIntegral usedBytes <= n) $ fail $ "Used more than " ++ show n ++ " bytes decoding the constant: " ++ show (prettyPlcDef t)
+                NoLimit -> pure ()
+            pure t
             where
                 -- Get the pointer where flat is currently decoding from. Requires digging into the innards of flat a bit.
                 getCurPtr :: Get (Ptr Word8)
                 getCurPtr = Get $ \_ s@S{currPtr} -> pure $ GetResult s currPtr
         handleTerm 5 = Force    <$> decode <*> go
         handleTerm 6 = Error    <$> decode
-        handleTerm 7 = Builtin  <$> decode <*> decode
+        handleTerm 7 = do
+            ann <- decode
+            fun <- decode
+            let t :: Term name uni fun ann
+                t = Builtin ann fun
+            unless (builtinPred fun) $ fail $ "Forbidden builtin function: " ++ show (prettyPlcDef t)
+            pure t
```

Comment:
Is there a reason you don't use return/break at tail position? like the comment suggestion above

Reply:
Not really, I didn't think it mattered much. I suspect they'll produce similar code.